provider/ns1.go

```diff
@@ -0,0 +1,295 @@
+	ns1Delete = "DELETE"
+	// ns1Update is a ChangeAction enum value
+	ns1Update = "UPDATE"
+	// ns1DefaultTTL is the default ttl for ttls that are not set
+	ns1DefaultTTL = 10
+)
+
+// NS1DomainClient is a subset of the NS1 API the the provider uses, to ease testing
+type NS1DomainClient interface {
+	CreateRecord(r *dns.Record) (*http.Response, error)
+	DeleteRecord(zone string, domain string, t string) (*http.Response, error)
+	UpdateRecord(r *dns.Record) (*http.Response, error)
+	GetZone(zone string) (*dns.Zone, *http.Response, error)
+	ListZones() ([]*dns.Zone, *http.Response, error)
+}
+
+// NS1DomainService wraps the API and fulfills the NS1DomainClient interface
+type NS1DomainService struct {
+	service *api.Client
+}
+
+// CreateRecord wraps the Create method of the API's Record service
+func (n NS1DomainService) CreateRecord(r *dns.Record) (*http.Response, error) {
+	return n.service.Records.Create(r)
+}
+
+// DeleteRecord wraps the Delete method of the API's Record service
+func (n NS1DomainService) DeleteRecord(zone string, domain string, t string) (*http.Response, error) {
+	return n.service.Records.Delete(zone, domain, t)
+}
+
+// UpdateRecord wraps the Update method of the API's Record service
+func (n NS1DomainService) UpdateRecord(r *dns.Record) (*http.Response, error) {
+	return n.service.Records.Update(r)
+}
+
+// GetZone wraps the Get method of the API's Zones service
+func (n NS1DomainService) GetZone(zone string) (*dns.Zone, *http.Response, error) {
+	return n.service.Zones.Get(zone)
+}
+
+// ListZones wraps the List method of the API's Zones service
+func (n NS1DomainService) ListZones() ([]*dns.Zone, *http.Response, error) {
+	return n.service.Zones.List()
+}
+
+// NS1Config passes cli args to the NS1Provider
+type NS1Config struct {
+	DomainFilter DomainFilter
+	ZoneIDFilter ZoneIDFilter
+	DryRun       bool
+}
+
+// NS1Provider is the NS1 provider
+type NS1Provider struct {
+	client       NS1DomainClient
+	domainFilter DomainFilter
+	zoneIDFilter ZoneIDFilter
+	dryRun       bool
+}
+
+// NewNS1Provider creates a new NS1 Provider
+func NewNS1Provider(config NS1Config) (*NS1Provider, error) {
+	return newNS1ProviderWithHTTPClient(config, http.DefaultClient)
+}
+
+func newNS1ProviderWithHTTPClient(config NS1Config, client *http.Client) (*NS1Provider, error) {
+	token, ok := os.LookupEnv("NS1_APIKEY")
+	if !ok {
+		return nil, fmt.Errorf("NS1_APIKEY environment variable is not set")
+	}
+
+	apiClient := api.NewClient(client, api.SetAPIKey(token))
+
+	provider := &NS1Provider{
+		client:       NS1DomainService{apiClient},
+		domainFilter: config.DomainFilter,
+		zoneIDFilter: config.ZoneIDFilter,
+	}
+	return provider, nil
+}
+
+// Records returns the endpoints this provider knows about
+func (p *NS1Provider) Records() ([]*endpoint.Endpoint, error) {
+	zones, err := p.zonesFiltered()
+	if err != nil {
+		return nil, err
+	}
+
+	var endpoints []*endpoint.Endpoint
+
+	for _, zone := range zones {
+
+		// TODO handle Header Codes
```

Comment:
What is missing here? Can we resolve this TODO instead of leaving it here?

Reply:
This TODO is to handle the edge case where NS1 paginates the response.  In this case this would only occur when a zone contains 10,000+ records.  Unfortunately, this is not yet handled by the NS1 GO SDK, so to resolve this now would require manually inspecting the headers and making additional requests for the next page of records.

If this project is OK with it, I'd prefer to open a PR against ns1-go to add pagination support and then come back to add support for this in the provider.